public/javascripts/common/SpeedLimit.js

```diff
@@ -0,0 +1,133 @@
+ * @param {function} isOnboarding Function that returns a boolean on whether the current mission is the tutorial/onboarding task
+ * @returns {SpeedLimit} SpeedLimit object with updateSpeedLimit function, container, speedLimit object with number and sub (units, e.g. 'mph'),
+ * speedLimitVisible boolean
+ */
+function SpeedLimit(panorama, coords, isOnboarding) {
+    let self = this;
+
+    function _init() {
+        self.container = document.getElementById("speed-limit-sign");
+        self.speedLimit = {
+            number: "",
+            sub: ""
+        };
+        self.speedLimitVisible = false
+        updateSpeedLimit();
+
+        // Listen for position changes.
+        panorama.addListener('position_changed', positionChange);
+    }
+
+    function updateSpeedLimit() {
+        self.container.querySelector("#speed-limit").innerText = self.speedLimit.number;
+        self.container.querySelector("#speed-limit-sub").innerText = self.speedLimit.sub;
+        self.container.style.display = self.speedLimitVisible ? "flex" : "none";
+    }
+
+    function findClosestRoad(data, lat, lon) {
+        // Filter to only be roads, and not foot paths/walk ways.
+        const roadHighwayTypes = [
+            "motorway",
+            "trunk",
+            "primary",
+            "secondary",
+            "tertiary",
+            "unclassified",
+            "residential",
+            "motorway_link",
+            "trunk_link",
+            "primary_link",
+            "secondary_link",
+            "tertiary_link",
+            "living_street",
+            "road"
+        ]
+        const roads = data.elements.filter(el => el.type === "way" && el.tags && el.tags.highway && roadHighwayTypes.includes(el.tags.highway));
+
+        const point = turf.point([lat, lon])
+        let closestRoad = null;
+        let minDistance = Infinity;
+
+        // Go through all the roads and find the closest one, keeping line segments in mind.
+        // Typical point comparison won't always work so it's better to compare to the segment itself.
+        for (const road of roads) {
+            const lineString = turf.lineString(road.geometry.map(p => [p.lon, p.lat]));
+            const distance = turf.pointToLineDistance(point, lineString);
+
+            if (distance < minDistance) {
+                minDistance = distance;
+                closestRoad = road;
+            }
+        }
+
+        return closestRoad;
+    }
+
+    async function positionChange() {
+        // If user is in the onboarding/tutorial mission, we can skip getting the speed limit and hide the sign altogether
+        if (isOnboarding()) {
+            self.speedLimitVisible = false
+            updateSpeedLimit()
+            return
+        }
+
+        // Get the current position.
+        const { lat, lng } = coords()
+
+        // Get nearby roads and their respective information from the overpass API.
+        const overpassQuery = `
+        [out:json];
```

Comment:
I haven't really worked directly with the Overpass API, is this typically how it's formatted? Looks kinda funky and confusing to me, but if that's how it's done that's fine!

Reply:
To be honest, I'm not quite sure. I have not used the overpass API before implementing this and I agree it looks weird lol. I think it matches with some of the examples they have [here](https://wiki.openstreetmap.org/wiki/Overpass_API/Overpass_API_by_Example).